### ISSUE TITLE
Add support for dynamic inputs `condition` key

### ISSUE THREAD

**Issue by ph** [elastic/package-spec#86]
In 7.10 we have added support for _condition_ , this key support adding a boolean expression to know if the specific part need to be executed.  In our [documentation](https://github.com/elastic/beats/blob/master/x-pack/elastic-agent/docs/elastic-agent-conditions.asciidoc#conditions) you can see you can apply that key in multiple places.

In this example, the input is applied if the host platform is Linux:

```yaml
inputs:
  - type: logfile
    streams:
      - paths:
         - /var/log/syslog
    condition: ${host.platform} == 'linux'
```

In this example, the stream is applied if the host platform is not Windows:

```yaml
inputs:
  - type: system/metrics
    streams:
      - metricset: load
        data_stream.dataset: system.cpu
        condition: ${host.platform} != 'windows'
```

In this example, the processor is applied if the host platform is not Windows:
```yaml
inputs:
  - type: system/metrics
    streams:
      - metricset: load
        data_stream.dataset: system.cpu
    processors:
      - add_fields:
          fields:
            platform: ${host.platform}
          to: host
        condition: ${host.platform} != 'windows'
```

We should add support in the spec for the `condition`

**Comment by ycombinator** [comment]
Hey @ph, thanks for the examples but those look like they are from an Agent configuration's POV. But since the package spec deals with package definitions, it would be useful to have the corresponding examples in terms of package definitions.

I already looked at https://github.com/elastic/integrations/pull/411 and found https://github.com/elastic/integrations/pull/411/files#diff-d4cd9d386b49496970c932d312ae09b5a2acc2c3f85f75a7819064d67634248bR39. I think this one corresponds to your first example ("the input is applied if the host platform is Linux"). Could you point me to or post similar _package_ definitions corresponding to your second and third examples? Thanks!

**Comment by ph** [comment]
@ycombinator You are correct, looking at this and the existing _integrations_ I do not see any specific usage of processors , this make sense because we want to push for ingest pipeline. so maybe we should just keep it at the input level as you have defined is enough and we can modify the spec if we ever need to?

WDYT @ruflin ?

**Comment by ruflin** [comment]
Condition can be used today on input and stream level I think. Same applies to processors. Would the goal of adding it to the spec be that we can validate it?

**Comment by ph** [comment]
@ruflin Yes, the current spec is strict if its not part of the manifest it's not authorized.

**Comment by ycombinator** [comment]
Given that the spec is strict but also GA (= breaking changes would need a major version bump), I'd suggest only adding fields to the spec that have a concrete need for today. 

When we bump into a problem with a new field or a new location for a field, we can always add it to the spec at that time. But adding fields we aren't actually using in any packages today is a bit risky if we need to then take them out later (breaking change) or if we just keep them around, thereby loosening the spec.

**Comment by ph** [comment]
@ycombinator I am ok with what you propose.

**Comment by ph** [comment]
looping @fearful-symmetry and @steffen on this.

**Comment by fearful-symmetry** [comment]
So, I'm trying to do something similar to this:
```
inputs:
  - type: logfile
    streams:
      - paths:
         - /var/log/syslog
    condition: ${host.platform} == 'linux'
```

And it doesn't appear to be working. Do we _expect_ this to work now?

**Comment by ycombinator** [comment]
> Do we expect this to work now?

No, not yet. I have a draft PR up for adding `condition` on an input definition in the package (https://github.com/elastic/package-spec/pull/87) but I was waiting on a clear direction (which has happened now: https://github.com/elastic/package-spec/issues/86#issuecomment-736805489) and I was also working on other tasks in the mean time.

I will convert my draft PR to fair now and put it into review. Once its merged and the change has been propagated to `elastic-package` (requiring another small dependency update PR), what you're trying to do [here](https://github.com/elastic/integrations/pull/411/files#diff-d4cd9d386b49496970c932d312ae09b5a2acc2c3f85f75a7819064d67634248bR39) should work.

**Comment by ph** [comment]
Sorry about that @fearful-symmetry I didn't follow up the issue correctly.

**Comment by ycombinator** [comment]
Nah, not on you @ph. I don't think I was clear enough earlier that I was waiting for input on this issue + I was getting pulled into other things so I didn't follow up here. Mea culpa 🖐️.

**Comment by ph** [comment]
:beers:

**Comment by ycombinator** [comment]
@fearful-symmetry You should be all set now with the `condition` change in https://github.com/elastic/integrations/pull/411/. 

If you're using `elastic-package` locally, you'll need to install the latest version. I've already updated the `elastic-package` dependency in the `integrations` repo, so your PR should be fine with CI.

**Comment by fearful-symmetry** [comment]
@ycombinator thanks a ton! I'll update everything and take it for a spin.

**Comment by nchaulet** [comment]
@ycombinator @ph Maybe I am misunderstanding something here, but could we use the agent input or stream template to specify condition?

**Comment by ycombinator** [comment]
@nchaulet could you point to a location in the package spec where the agent input and stream template are specified in a package? That might help me understand what you're suggesting. Thanks!

**Comment by nchaulet** [comment]
@ycombinator I was thinking of using the template path as defined here  https://github.com/elastic/package-spec/blob/master/versions/1/manifest.spec.yml#L210

the manifest file will be something like this
```yaml
policy_templates:
  inputs:
    - type: logfile
      template_path: winlog.hbs.yml
      streams:
        - paths:
           - /var/log/syslog
```

and in /agent/input/winlog.hbs.yml
```yaml
condition: ${host.platform} == 'windows'
```

And this will generate the following policy for the agent

```taml
inputs:
  - type: logfile
    streams:
      - paths:
         - /var/log/syslog
    condition: ${host.platform} == 'windows'
```

**Comment by ycombinator** [comment]
Understood, thanks for that @nchaulet. What you are suggesting definitely works!

I think then it comes down to whether we expect **most** packages to specify a `condition` for their inputs or not.
- If the answer is "yes", then promoting `condition` to be a direct property of an input (as we did in https://github.com/elastic/package-spec/pull/87) would make sense.
- If the answer is "no", then requiring package authors to specify the condition via the `template_path` mechanism you suggest is okay. Its a bit more indirect but that's okay if most package authors don't need to use it.

I think the answer is going to be "no", considering that none of the existing 34 packages in https://github.com/elastic/integrations have needed `condition` on their 65 inputs so far. So personally I'm okay with reverting the change in https://github.com/elastic/package-spec/pull/87 and suggesting your approach for https://github.com/elastic/integrations/pull/411. @mtojek @ruflin @ph WDYT?

**Comment by ycombinator** [comment]
@fearful-symmetry Heads up! You might want to hold on merging https://github.com/elastic/integrations/pull/411 until the discussion in the past four comments above is resolved. I expect we'll have a resolution on this within a day or so.

**Comment by ph** [comment]
Okay, sorry I've read the issue again, thanks @nchaulet and @ruflin for pointing that I was wrong. I think what @nchaulet proposition is making more sense to what I've been pushing having a condition to be a direct property of the input even if the Agent supports it.

So I think going with **no**, and having author use the `template_path` mechanism is an OK and it also reduce future changes in spec and in kibana.

So I would be to +1 to revert #87, use the [proposition](https://github.com/elastic/package-spec/issues/86#issuecomment-740352800) as the official way to suppor this.

**Comment by ph** [comment]
@fearful-symmetry can you verify the above for both the windows and the non windows case and close this issue when validated?

**Comment by fearful-symmetry** [comment]
@ph Alright, let me make sure I have my head wrapped around this. Is @nchaulet's implementation of the conditional in the `stream.hbs.yml` working _now_ ?

**Comment by ycombinator** [comment]
PR to revert #87 is up: https://github.com/elastic/package-spec/pull/93

**Comment by nchaulet** [comment]
@fearful-symmetry Yes the conditional in stream should be working now

**Comment by ph** [comment]
I am going to close this as wont fix.

@ycombinator @mtojek Is that documentated in the pacakge spec?
